agent/src/apiclient.rs

```diff
@@ -129,27 +129,41 @@ async fn invoke_apiclient(args: Vec<String>) -> Result<Output> {
             .output()
             .context(apiclient_error::ApiClientRawCommand { args: args.clone() })?;
 
-        if output.status.success() {
-            return Ok(output);
-        }
-        let error_content = String::from_utf8_lossy(&output.stderr).to_string();
-        let error_statuscode = extract_status_code_from_error(&error_content);
-
-        match error_statuscode {
-            UPDATE_API_BUSY_STATUSCODE => {
-                event!(Level::INFO, "API server busy, retrying later ...");
-                // Retry after ten seconds if we get a 423 Locked response (update API busy)
-                sleep(UPDATE_API_SLEEP_DURATION).await;
-                attempts += 1;
-            }
-            _ => {
-                // API response was a non-transient error, bail out
-                return apiclient_error::BadHttpResponse {
-                    statuscode: error_statuscode,
+        match output.status.success() {
+            true => return Ok(output),
+            false => {
+                match output.status.code() {
+                    Some(_code) => {
+                        let error_content = String::from_utf8_lossy(&output.stderr).to_string();
+                        let error_statuscode = extract_status_code_from_error(&error_content);
+
+                        match error_statuscode {
+                            UPDATE_API_BUSY_STATUSCODE => {
+                                event!(Level::INFO, "API server busy, retrying later ...");
+                                // Retry after ten seconds if we get a 423 Locked response (update API busy)
+                                sleep(UPDATE_API_SLEEP_DURATION).await;
+                                attempts += 1;
+                            }
+                            _ => {
+                                // API response was a non-transient error, bail out
+                                return apiclient_error::BadHttpResponse {
+                                    statuscode: error_statuscode,
+                                }
+                                .fail();
+                            }
+                        };
+                    }
+                    // reboot sends signal to kill the agent; instead of that, we should gracefully exit brupop agent.
+                    _ => {
+                        event!(
```

Comment:
I'm a bit surprised that we can handle reboots this deep in the abstraction layer, and not higher up when we actually call the API to trigger a reboot.

I think more specifically, I'm not sure I understand how a reboot ultimately causes this codepath to be executed. Can you elaborate?

Reply:
Yeah! When I looked into the error `thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1'` and code, I found the root cause to make agent panic is [here](https://github.com/bottlerocket-os/bottlerocket-update-operator/blob/develop/agent/src/apiclient.rs#L112).

When apiclient successfully called `reboot`, I expected it to return `output.status.successed() == true`. However, it returned false and stderr was empty even the `reboot` worked as expected. The status of reboot command was `Exit Status(unix wait status(15)), stderr=''` which means a request to the program to terminate. Therefore, I think I should consider this situation as success when calls `reboot`. Add an extra match condition there to handle the signal which sends to kill the agent, and gracefully exit the agent.

Reply:
Would it make sense to only `exit()` if we receive that exist status? It's hard reading this code to understand the interaction you mentioned here.

Reply:
you mean it's bette to do like this?  Exist on specific Exit status?
```
match error_statuscode {
... =>....,
15 => std::process::exit(0)
}
```
The reason why I didn't do that was that the logic to retrieve the exit status [here](https://github.com/bottlerocket-os/bottlerocket-update-operator/blob/develop/agent/src/apiclient.rs#L135-L136) is to extract exit status from stderr. However, the status of calling `reboot` is `fail` but `stderr` is empty so that unable to extract the exist status. Therefore, I have the logic here if the command returns fail but with empty stderr, It should be reboot and we need exit  agent.

Yeah, I agree that's not a straightforward way and inconvenient to read. I'll think about a new way which can make agent exist on the termination signal.

Reply:
You mentioned:

> The status of reboot command was Exit Status(unix wait status(15)), stderr=''

Are we saying that the `apiserver` received signal 15 (SIGTERM) and still returned this result? Or did our brupop agent receive SIGTERM? If it's our own agent, we could possibly write a [signal handler](https://stackoverflow.com/questions/26280859/how-to-catch-signals-in-rust). If it's from the apiserver, can we extract it from the Exit Status you mentioned above somehow?

I'd be okay with explaining the situation in a comment if there's not a better way to make it more clear in the code.

Reply:
Yeah, `apiserver` received signal 15 (SIGTERM), and I was trying to extract Exit Status from the stderr , but it was empty there. It looks like signal 15 (SIGTERM) isn't triggered by any errors but it acts like being triggered by errors.

After further investment, I would say current method is more clear. I'll add more explanation to make code more readable.

Reply:
Yeah, `apiserver` received signal 15 (SIGTERM). 

On Unix, this will return `None` if the process was terminated by a signal (reboot sends signal 15 (SIGTERM)). Signal termination is not considered a success. After further investment, I would say current method is more clear. I'll add more explanation to make code more readable.